File: apps/ios/Packages/Sources/ShadowRoomTimelineFeature/RoomTimelineViewModel.swift
import Combine
import Foundation

@MainActor
public final class RoomTimelineViewModel: ObservableObject {
    @Published public private(set) var state: RoomTimelineState
    @Published public private(set) var securityState: RoomTimelineSecurityState = .unknown
    @Published public private(set) var draft = ""
    @Published public private(set) var isSending = false
    @Published public private(set) var sendErrorMessage: String?

    private let roomId: String
    private let repository: RoomTimelineRepository

    public init(
        roomId: String,
        repository: RoomTimelineRepository,
        initialState: RoomTimelineState = .loading
    ) {
        self.roomId = roomId
        self.repository = repository
        self.state = initialState
    }

    public func send(_ event: RoomTimelineEvent) {
        switch event {
        case .refreshRequested, .retryRequested:
            Task { await load() }
        case .draftChanged(let draft):
            self.draft = draft
            sendErrorMessage = nil
        case .sendRequested:
            Task { await sendDraft() }
        }
    }

    public func load() async {
        state = .loading

        do {
            apply(try await repository.loadTimeline(roomId: roomId))
        } catch {
            state = .failed
        }
    }

    public func observeTimeline() async {
        state = .loading

        do {
            apply(try await repository.loadTimeline(roomId: roomId))
            let updates = try await repository.timelineUpdates(roomId: roomId)
            for await snapshot in updates {
                try Task.checkCancellation()
                apply(snapshot)
            }
        } catch is CancellationError {
            return
        } catch {
            guard !Task.isCancelled else { return }
            state = .failed
        }
    }

    public func sendDraft() async {
        guard !isSending else { return }
        let body = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !body.isEmpty else { return }

        switch state {
        case .loaded, .empty:
            break
        case .loading, .failed:
            sendErrorMessage = "Nachrichten können erst nach dem Laden des Raums gesendet werden."
            return
        }

        isSending = true
        sendErrorMessage = nil
        defer { isSending = false }

        do {
            try await repository.sendMessage(roomId: roomId, body: body)
            draft = ""
        } catch RoomTimelineRepositoryError.sendingUnavailable {
            sendErrorMessage = "Für diesen Raum ist noch keine Send-Pipeline verbunden."
        } catch {
            sendErrorMessage = "Die Nachricht konnte nicht gesendet werden."
        }
    }

    private func apply(_ snapshot: RoomTimelineSnapshotViewState) {
        securityState = snapshot.securityState
        if snapshot.items.isEmpty {
            state = .empty(roomTitle: snapshot.roomTitle)
        } else {
            state = .loaded(roomTitle: snapshot.roomTitle, items: snapshot.items)
        }
    }
}

File: apps/android/features/timeline/src/main/java/com/shadowchat/features/timeline/RoomTimelineViewModel.kt
package com.shadowchat.features.timeline

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class RoomTimelineViewModel(
    private val roomId: String,
    private val repository: RoomTimelineRepository,
) : ViewModel() {
    private val _state = MutableStateFlow<RoomTimelineUiState>(RoomTimelineUiState.Loading)
    val state: StateFlow<RoomTimelineUiState> = _state.asStateFlow()

    fun onEvent(event: RoomTimelineEvent) {
        when (event) {
            RoomTimelineEvent.Appeared -> {
                if (_state.value == RoomTimelineUiState.Loading) {
                    load()
                }
            }
            RoomTimelineEvent.RefreshRequested,
            RoomTimelineEvent.RetryRequested -> load()
        }
    }

    fun load() {
        _state.value = RoomTimelineUiState.Loading

        viewModelScope.launch {
            _state.value = try {
                val snapshot = repository.loadTimeline(roomId)
                if (snapshot.items.isEmpty()) {
                    RoomTimelineUiState.Empty(roomTitle = snapshot.roomTitle)
                } else {
                    RoomTimelineUiState.Loaded(
                        roomTitle = snapshot.roomTitle,
                        items = snapshot.items,
                    )
                }
            } catch (_: Throwable) {
                RoomTimelineUiState.Error
            }
        }
    }
}
